server/service/src/localisations.rs

```diff
@@ -49,39 +75,94 @@ impl Localisations {
 
     // Get a translation for a given key and language
     // next need to add fallback and namespace to get Translation function
-    pub fn get_translation(&self, key: &str, language: &str, namespace: &str, fallback: &str ) -> String {
-        self.translations
-            .get(language)
-            .and_then(|map| map.get(namespace))
-            .and_then(|map| map.get(key))
-            .cloned()
-            .unwrap_or_else(|| fallback.to_string())
+    pub fn get_translation(&self, args: &HashMap<String, serde_json::Value> ) -> Result<String, TranslationError> {
+        let key = args.get("key").and_then(serde_json::Value::as_str).unwrap_or("").to_string();
+        let lang = args.get("lang").and_then(serde_json::Value::as_str).unwrap_or("en").to_string();
+        // use common.json namespace if nominated namespace can't be found
+        let namespace = args.get("namespace").and_then(serde_json::Value::as_str).unwrap_or("common.json").to_string();
+        let fallback = args.get("fallback").and_then(serde_json::Value::as_str);
+        match self.translations
+        .get(&lang)
+        .and_then(|map| map.get(&namespace))
+        .and_then(|map| map.get(&key))
+        .cloned() {
+            Some(translation) => Ok(translation),
+            None => {
+                // Fall back to translation from common.json if key doesn't exist in nominated namespace.
```

Comment:
I was thinking after doing this that this fallback structure isn't perfect because as a result of the nested hashmap structure, it won't differentiate between the translation not existing in the file, or the file not existing in the dir.  

So there might be gaps here?

It seems relatively rigorous to me though in that it will pull a translation or it will fallback, or it will fail loudly.